internal/providers/virtualbox/virtualbox.c

```diff
@@ -0,0 +1,206 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <linux/vboxguest.h>
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include "virtualbox.h"
+
+static void _cleanup_close(int *fd) {
+	if (*fd != -1) {
+		close(*fd);
+	}
+}
+#define _cleanup_close_ __attribute__((cleanup(_cleanup_close)))
+
+static void _cleanup_free(void *ptr) {
+	free(*(void **)ptr);
+}
+#define _cleanup_free_ __attribute__((cleanup(_cleanup_free)))
+
+static void init_header(struct vbg_ioctl_hdr *hdr, size_t size_in, size_t size_out) {
+	hdr->size_in = sizeof(*hdr) + size_in;
+	hdr->version = VBG_IOCTL_HDR_VERSION;
+	hdr->type = VBG_IOCTL_HDR_TYPE_DEFAULT;
+	hdr->size_out = sizeof(*hdr) + size_out;
+}
+
+static int version_info(int fd) {
+	struct vbg_ioctl_driver_version_info msg = {
+		.u = {
+			.in = {
+				.req_version = 0x00010000,
+				.min_version = 0x00010000,
+			},
+		},
+	};
+	init_header(&msg.hdr, sizeof(msg.u.in), sizeof(msg.u.out));
+	if (ioctl(fd, VBG_IOCTL_DRIVER_VERSION_INFO, &msg)) {
+		return VERR_GENERAL_FAILURE;
+	}
+	return msg.hdr.rc;
+}
+
+static int connect(int fd, uint32_t *client_id) {
+	struct vbg_ioctl_hgcm_connect msg = {
+		.u = {
+			.in = {
+				.loc = {
+					.type = VMMDEV_HGCM_LOC_LOCALHOST_EXISTING,
+					.u = {
+						.localhost = {
+							.service_name = "VBoxGuestPropSvc",
+						},
+					},
+				},
+			},
+		},
+	};
+	init_header(&msg.hdr, sizeof(msg.u.in), sizeof(msg.u.out));
+	if (ioctl(fd, VBG_IOCTL_HGCM_CONNECT, &msg)) {
+		return VERR_GENERAL_FAILURE;
+	}
+	if (msg.hdr.rc != VINF_SUCCESS) {
+		return msg.hdr.rc;
+	}
+	*client_id = msg.u.out.client_id;
+	return VINF_SUCCESS;
+}
+
+static int get_prop(int fd, uint32_t client_id, const char *name, void **value, size_t *size) {
+	// init header
+	size_t msg_size = sizeof(struct vbg_ioctl_hgcm_call) + 4 * sizeof(struct vmmdev_hgcm_function_parameter64);
+	struct vbg_ioctl_hgcm_call _cleanup_free_ *msg = calloc(1, msg_size);
+	// init_header re-adds the size of msg->hdr
+	init_header(&msg->hdr, msg_size - sizeof(msg->hdr), msg_size - sizeof(msg->hdr));
+	msg->client_id = client_id;
+	msg->function = 1;     // GUEST_PROP_FN_GET_PROP
+	msg->timeout_ms = -1;  // inf
+	msg->interruptible = 1;
+	msg->parm_count = 4;
+
+	// init arguments
+	char ch;
+	struct vmmdev_hgcm_function_parameter64 *params = (void *) (msg + 1);
+	// property name (in)
+	params[0].type = VMMDEV_HGCM_PARM_TYPE_LINADDR_IN;
+	params[0].u.pointer.size = strlen(name) + 1;
+	params[0].u.pointer.u.linear_addr = (uintptr_t) name;
+	// property value (out)
+	params[1].type = VMMDEV_HGCM_PARM_TYPE_LINADDR;
+	params[1].u.pointer.size = 1;
+	params[1].u.pointer.u.linear_addr = (uintptr_t) &ch;
+	// property timestamp (out)
+	params[2].type = VMMDEV_HGCM_PARM_TYPE_64BIT;
+	// property size (out)
```

Comment:
Is this a "bytes count" or a "string length" size?
Depending on the answer, I think there may be either a `malloc(0)` corner-case to handle or an off-by-one in the `C.GoBytes` conversion.